src/Lean/Server/Watchdog.lean

```diff
@@ -376,6 +376,33 @@ def handleWorkspaceSymbol (p : WorkspaceSymbolParams) : ServerM (Array SymbolInf
     |>.map fun ((name, _), location) =>
       { name, kind := SymbolKind.constant, location }
 
+def handlePrepareRename (p : TextDocumentPositionParams) : ServerM (Option Range) := do
+  -- This just checks that the cursor is over a renameable identifier
+  if let some path := System.Uri.fileUriToPath? p.textDocument.uri then
+    let srcSearchPath := (← read).srcSearchPath
+    if let some module ← searchModuleNameOfFileName path srcSearchPath then
+      let references ← (← read).references.get
+      return references.findRange? module p.position (includeStop := true)
+  return none
+
+def handleRename (p : RenameParams) : ServerM Lsp.WorkspaceEdit := do
+  if (String.toName p.newName).isAnonymous then
+    throwServerError s!"Can't rename: `{p.newName}` is not an identifier"
+  let mut refs : HashMap DocumentUri (RBMap Lsp.Position Lsp.Position compare) := ∅
+  for { uri, range } in (← handleReference { p with context.includeDeclaration := true }) do
+    refs := refs.insert uri <| (refs.findD uri ∅).insert range.start range.end
+  -- We have to filter the list of changes to put the ranges in order and
+  -- remove any duplicates or overlapping ranges, or else the rename will not apply
+  let changes := refs.fold (init := ∅) fun changes uri map => Id.run do
+    let mut last := ⟨0, 0⟩
+    let mut arr := #[]
+    for (start, stop) in map do
+      if last ≤ start then
+        arr := arr.push { range := ⟨start, stop⟩, newText := p.newName }
+        last := stop
+    return changes.insert uri arr
+  return { changes? := some changes }
```

Comment:
Issues 3, 4, and 5 are expected and I don't think they should be addressed in this PR, as that would require a major overhaul to the data stored in .ilean files. Even though this mechanism is imprecise and likely requires manual fixup, it is still way better than the status quo of nothing at all. We can iterate on improving it later.

Issues 1 and 2 sound like bugs in the references data structure, I will investigate.

Reply:
Can we turn these examples into tests?

Reply:
As mentioned in the OP, tests for this feature don't work in the lean test suite because rename doesn't work on bare files and all of the `interactive/` tests are bare files (and the `pkg/` tests don't have the tooling to support interactive tests). I'm testing them manually. But I can add them to the test suite anyway.

Reply:
@mhuisi I've tracked down the issue for 1. When you write `structure Foo where` without `mk ::`, lean uses `Foo` as the position for the `Foo.mk` declaration, which makes sense if you want to handle go-to-def on uses of `Foo.mk`. But we need to indicate that this is not a "canonical" syntax for the identifier so that rename on `Foo` is not interpreted as renaming `Foo.mk` (I think there are multiple renames on the same location and it just picks a random one in this case). I am inclined to fix this in a follow up PR, as this will require changes to the .ilean data structure to handle correctly.